src/modules/cmdpal/Microsoft.CmdPal.UI.ViewModels/ListViewModel.cs

```diff
@@ -31,80 +34,137 @@ public partial class ListViewModel : PageViewModel
+
     public ListViewModel(IListPage model, TaskScheduler scheduler)
         : base(model, scheduler)
     {
         _model = new(model);
     }
 
+    private void Model_ItemsChanged(object sender, ItemsChangedEventArgs args) => FetchItems();
+
     protected override void OnFilterUpdated(string filter)
     {
         //// TODO: Just temp testing, need to think about where we want to filter, as ACVS in View could be done, but then grouping need CVS, maybe we do grouping in view
         //// and manage filtering below, but we should be smarter about this and understand caching and other requirements...
         //// Investigate if we re-use src\modules\cmdpal\extensionsdk\Microsoft.CmdPal.Extensions.Helpers\ListHelpers.cs InPlaceUpdateList and FilterList?
 
-        // Remove all items out right if we clear the filter, otherwise, recheck the items already displayed.
-        if (string.IsNullOrWhiteSpace(filter))
+        // Dynamic pages will handler their own filtering. They will tell us if
+        // something needs to change, by raising ItemsChanged.
+        if (_isDynamic)
         {
-            Items.Clear();
-        }
-        else
-        {
-            // Remove any existing items which don't match the filter
-            for (var i = Items.Count - 1; i >= 0; i--)
+            try
             {
-                if (!Items[i].MatchesFilter(filter))
+                if (_model.Unsafe is IDynamicListPage dynamic)
                 {
-                    Items.RemoveAt(i);
+                    dynamic.SearchText = filter;
                 }
             }
-        }
-
-        // Add back any new items which do match the filter
-        foreach (var item in _itemCache)
-        {
-            if ((filter == string.Empty || item.MatchesFilter(filter))
-                && !Items.Contains(item)) //// TODO: We should be smarter here somehow
+            catch (Exception ex)
             {
-                Items.Add(item);
+                ShowException(ex);
             }
         }
+        else
+        {
+            // But for all normal pages, we should run our fuzzy match on them.
+            ApplyFilter();
+        }
     }
 
-    private void Model_ItemsChanged(object sender, ItemsChangedEventArgs args) => FetchItems();
-
     //// Run on background thread, from InitializeAsync or Model_ItemsChanged
     private void FetchItems()
     {
         // TEMPORARY: just plop all the items into a single group
         // see 9806fe5d8 for the last commit that had this with sections
-        // TODO unsafe
         try
         {
             var newItems = _model.Unsafe!.GetItems();
 
+            // Collect all the items into new viewmodels
+            Collection<ListItemViewModel> newViewModels = [];
+
+            // TODO we can probably further optimize this by also keeping a
+            // HashSet of every ExtensionObject we currently have, and only
+            // building new viewmodels for the ones we haven't already built.
```

Comment:
Should we just make both our collections ObservableHashSets now and slurp this in? RE: https://github.com/zadjii-msft/PowerToys/pull/214#issuecomment-2537450436

> I did find too that EF core has an ObservableHashSet here: https://github.com/dotnet/efcore/blob/main/src/EFCore/ChangeTracking/ObservableHashSet.cs - so that seems like a good idea for the ListViewModel.Items to expediate the dup checks too (assuming we keep this setup). I assume we'd just add to the Notice.md if we copied it over (as I don't think it'd be easily consumable from NuGet). Also, assuming @Sergio0694 doesn't want it in the MVVM Toolkit's collections, maybe we could open a .NET issue for inclusion in the SDK directly in the future?

Reply:
Same thought here about the ordering. 

And IMO if an app _really_ wants to give us duplicates in the list, I think I'm fine with that?